rfcs/2021-05-23-broadcast.md

```diff
@@ -0,0 +1,50 @@
+# SMP agent broadcast
+
+## Problem
+
+Support agent message broadcast to multiple connections.
+
+It is done in ad-hoc way as part of the previous [groups proposal](./2021-05-15-groups.md) - this proposal defines broadcast as a separate agent primitive to simplify group management.
+
+It can also be used for other purposes when the same message needs to be sent to multiple recipients without creating groups.
+
+## Solution
+
+A minimal protocol of additional client commands to create, manage and use broadcasts.
+
+From the point of view of the recipient this will look like a normal message, as if the sending agent executed multiple send commands (in fact, broadcast can be implemented by agent sending itself multiple SEND commands)
+
+### Commands and messages
+
+- command `B:bId? NEW` - create broadcast (response is `B:bId OK`, or `ERR` if broadcast already exists)
+- command `B:bId ADD C:cId` - add existing connection to a broadcast (response is `B:bId OK` or `ERR`, e.g. if connection already added or does not exist)
+- command `B:bId SEND msg` - broadcast message (response is multiple `C:cId SENT msgId` or ERR, separately for each connection)
+- message `B:bId SENT [C:bId] msgId` - notification that the message is sent to a specific or all recipients
```

Comment:
so is response `C:cId SENT msgId` or `B:bId SENT [C:bId] msgId`? feels like these are the same and should be aligned, right? or is the second notification after all per connection responses?

Reply:
yes, didn't settle yet on what is better.

In the current implementation it is `C:cId SENT msgId`, but probably it is better to make it `B:bId SENT C:bId msgId` and then once sent to all `B:bId SENT msgId`. Although maybe not :)

There definitely should be notification per connection and then once all delivered per broadcast.

That, by the way, requires that the agent client, when it is coded, understands that one command can have multiple responses with the same correlation ID and can collect them all prior to returning the function result - SMP protocol does not have this problem - for command responses that do not depend on other agents and only depend on the servers.